File: src/main/java/com/vc/medium/L959.scala
package com.vc.medium

object L959 {
  def regionsBySlashes(grid: Array[String]): Int = {
    val n = grid.length
    val m = grid(0).length
    val nNew = n * 3
    val mNew = m * 3

    val g = Array.ofDim[Int](nNew, mNew)
    grid.indices.foreach(i => {
      grid(i).indices.foreach(j => {
        if(grid(i)(j) == '/') {
          g(i * 3)(j * 3 + 2) = 1
          g(i * 3 + 1)(j * 3 + 1) = 1
          g(i * 3 + 2)(j * 3) = 1
        }
        else if(grid(i)(j) == '\\') {
          g(i * 3)(j * 3) = 1
          g(i * 3 + 1)(j * 3 + 1) = 1
          g(i * 3 + 2)(j * 3 + 2) = 1
        }
      })
    })

    val dirs = Array((1, 0), (0, 1), (-1, 0), (0, -1))
    def dfs(x: Int, y: Int): Unit = {
      if(g(x)(y) == 0) {
        g(x)(y) = 1
        dirs.foreach(dir => {
          val xNew = x + dir._1
          val yNew = y + dir._2
          if(xNew >= 0 && xNew < nNew && yNew >= 0 && yNew < mNew) {
            dfs(xNew, yNew)
          }
        })
      }
    }

    var res = 0
    g.indices.foreach(i => {
      g(i).indices.foreach(j => {
        if(g(i)(j) == 0) {
          dfs(i, j)
          res += 1
        }
      })
    })
    res
  }
}

File: src/main/java/com/vc/medium/L959.java
package com.vc.medium;

class L959 {
    public int regionsBySlashes(String[] grid) {
        /*
            Represent each square as 3 * 3 matrix
            [" /"]
            ["/ "]

            one represent a line and then problem reduces to find number of island
            [

             1st Col    2nd Col
              0 0 0      0 0 1
              0 0 0      0 1 0   1st Row
              0 0 0      1 0 0

              0 0 1      0 0 0
              0 1 0      0 0 0   2nd Row
              1 0 0      0 0 0
            ]
        */

        if(grid.length == 0) return 0;

        int n = grid.length;
        int m = grid[0].length();

        int[][] matrix = new int[3 * n][3 * m];
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                if(grid[i].charAt(j) == '/') {
                    matrix[i * 3][j * 3 + 2] = 1;
                    matrix[i * 3 + 1][j * 3 + 1] = 1;
                    matrix[i * 3 + 2][j * 3] = 1;
                }
                else if(grid[i].charAt(j) == '\\') {
                    matrix[i * 3][j * 3] = 1;
                    matrix[i * 3 + 1][j * 3 + 1] = 1;
                    matrix[i * 3 + 2][j * 3 + 2] = 1;
                }
            }
        }

        // for(int i = 0; i < matrix.length; i++) {
        //     for(int j = 0; j < matrix[0].length; j++) {
        //         System.out.print(matrix[i][j]+" ");
        //     }
        //     System.out.println("");
        // }

        return distinctIsland(matrix);
    }

    private int distinctIsland(int[][] matrix) {
        int count = 0;
        for(int i = 0; i < matrix.length; i++) {
            for(int j = 0; j < matrix[i].length; j++) {
                if(matrix[i][j] == 0) {
                    count++;
                    dfs(i, j, matrix);
                }
            }
        }
        return count;
    }

    private int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    private void dfs(int row, int col, int[][] matrix) {
        if(matrix[row][col] == 1) return;

        matrix[row][col] = 1;
        for(int[] dir: dirs) {
            int rowNew = dir[0] + row;
            int colNew = dir[1] + col;
            if(rowNew >= 0 && rowNew < matrix.length && colNew >= 0 && colNew < matrix[0].length) {
                dfs(rowNew, colNew, matrix);
            }
        }
    }
}
